File: AtCoder/ABC/ABC101-150/abc116/abc116d.py
n, k = list(map(int, input().split()))
tk = [list(map(int, input().split())) for i in range(n)]
ans = 0
tk.sort(key=lambda x: -x[1])
max_ = [[i, 0, 0] for i in range(n+1)]
for i in range(n):
  if max_[tk[i][0]][1]<tk[i][1]:
    max_[tk[i][0]][1], max_[tk[i][0]][2] = tk[i][1], i
max_.sort(key=lambda x: -x[1])
for i in range(1, k+1):
  neta = set()
  have = set()
  cnt = 0
  cur = 0
  for j in range(i):
    if max_[j][1]>0:
      cur += max_[j][1]
      cnt += 1
      neta.add(max_[j][0])
      have.add(max_[j][2])
  for j in range(n):
    if j in have:
      continue
    if cnt==k:
      break
    if len(neta)<i:
      neta.add(tk[j][0])
      cnt += 1
      cur += tk[j][1]
    elif len(neta)==i and tk[j][0] in neta:
      cnt += 1
      cur += tk[j][1]
  ans = max(ans, cur+len(neta)**2)
print(ans)

File: AtCoder/ABC/ABC101-150/abc116/abc116d.cpp
#include <bits/stdc++.h>
using namespace std;
#define REP(i,b,e) for(int i=(b);i<(e);++i)
#define RREP(i,b,e) for(int i=(b)-1;i>=e;--i)
#define rep(i,e) for(int i=0;i<(e);++i)

inline void print(void) { cout<<'\n'; }
template<class T> inline void print(const T &x) { cout<<x<<'\n'; }
template<class T, class... U> inline void print(const T &x, const U&... y) { cout<<x<<" "; print(y...); }

int main() {
  int n, k; cin>>n>>k;
  vector<pair<int, int>> td(n);
  rep(i, n) cin>>td[i].first>>td[i].second;
  sort(td.begin(), td.end(), [](auto x, auto y) { return x.second > y.second; });

  long long ans = 0, cur = 0;
  set<int> s;
  stack<int> q;
  rep(i, k) {
    if (s.count(td[i].first)) q.push(td[i].second);
    s.emplace(td[i].first);
    cur += td[i].second;
  }
  ans = cur + s.size() * s.size();

  REP(i, k, n) {
    if (q.empty()) break;
    if (s.count(td[i].first)) continue;
    cur -= q.top(); q.pop();
    cur += td[i].second;
    s.emplace(td[i].first);
    ans = max(ans, cur + (long long)(s.size() * s.size()));
  }
  print(ans);

  return 0;
}